[Index Management] Persist existing data_stream_options in component and index templates

SoniaSanzV opened the pull request:
Closes https://github.com/elastic/kibana/issues/217115

## Summary
Elasticsearch has introduced a new field to component and index templates called data_stream_options that enables failure stores on a data stream. This field is not available for legacy templates. This PR ensure that `data_stream_options` values are persisted in the component/index template creation wizards, if previously set via the API.

We will not surface the ability to change the configuration via the UI until we have time to discuss the proper user experience.

### How to test
1. Create a index template via console:
```
PUT _index_template/failure-store-index-template
{
  "index_patterns": ["failurestore-test-*"],
  "data_stream": {},
  "template": {
    "data_stream_options": {
      "failure_store": {
        "enabled": true
      }
    }
  }
}
```

2. Edit the template with the UI and save it.
3. Come back to console and verify the `data_stream_options` are persisted in the template after the changes `GET _index_template/failure-store-index-template`

Repeat the process for a component template:
```
PUT _component_template/failure-store-component-template
{
  "template": {
    "data_stream_options": {
      "failure_store": {
        "enabled": true
      }
    }
  }
}
```



<!--ONMERGE {"backportTargets":["8.19"]} ONMERGE-->

SoniaSanzV committed:
Persist existing data_stream_options in component and index templates

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/298915)
* Commit: 8fecca704cec7e77a248e486d408d3e3fdf3c40b

### Failed CI Steps
* [Check Types](https://buildkite.com/elastic/kibana-pull-request/builds/298915#0196a4c3-0fa4-4a05-8dc1-7ab1f218fe27)

cc @SoniaSanzV

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/298915","number":298915,"commit":"8fecca704cec7e77a248e486d408d3e3fdf3c40b"}],"number":298915}
buildkite-pr-comment-->

SoniaSanzV committed:
handle index template in serialization

SoniaSanzV committed:
Fix linter

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/298961)
* Commit: dffba95833e0599831ba063560dbcfb4bd8bcc40

### Failed CI Steps
* [FTR Configs #8](https://buildkite.com/elastic/kibana-pull-request/builds/298961#0196a52c-a1c1-4f94-94cb-276ad2adc1d7)
* [FTR Configs #22](https://buildkite.com/elastic/kibana-pull-request/builds/298961#0196a52c-a1d2-4123-9c0f-bb0a9f56733b)
* [FTR Configs #39](https://buildkite.com/elastic/kibana-pull-request/builds/298961#0196a52c-a1e5-4170-84e6-953cea56e45d)
* [FTR Configs #39](https://buildkite.com/elastic/kibana-pull-request/builds/298961#0196a548-f8cc-4800-accd-445b3e047b1b)
* [FTR Configs #82](https://buildkite.com/elastic/kibana-pull-request/builds/298961#0196a52c-a205-4e07-b8e9-6f285ea6809d)
* [FTR Configs #93](https://buildkite.com/elastic/kibana-pull-request/builds/298961#0196a52c-a20d-4c5c-ad7d-cbb97865b4ae)
* [FTR Configs #121](https://buildkite.com/elastic/kibana-pull-request/builds/298961#0196a52c-a222-467c-988f-24e8dfb4cc5d)
* [FTR Configs #126](https://buildkite.com/elastic/kibana-pull-request/builds/298961#0196a52c-a226-45da-8bfd-deb580b80a5c)

### Test Failures
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/298961#0196a52c-a1e5-4170-84e6-953cea56e45d) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/298961/jobs/0196a52c-a1e5-4170-84e6-953cea56e45d/artifacts/0196a548-d3e6-468c-9684-834ecb1846fb) FTR Configs #<span></span>39 / Serverless Common UI - Management Index Management Index Templates Create index template Creates index template
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/298961#0196a52c-a205-4e07-b8e9-6f285ea6809d) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/298961/jobs/0196a52c-a205-4e07-b8e9-6f285ea6809d/artifacts/0196a54a-b2e5-427b-9917-7ac5e48de4db) FTR Configs #<span></span>82 / Serverless Common UI - Management Index Management Index Templates Create index template Creates index template
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/298961#0196a52c-a20d-4c5c-ad7d-cbb97865b4ae) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/298961/jobs/0196a52c-a20d-4c5c-ad7d-cbb97865b4ae/artifacts/0196a54a-5531-4efd-908a-b8d4504cfc1c) FTR Configs #<span></span>93 / Serverless Common UI - Management Index Management Index Templates Create index template Creates index template
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/298961#0196a548-f8cc-4800-accd-445b3e047b1b) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/298961/jobs/0196a548-f8cc-4800-accd-445b3e047b1b/artifacts/0196a554-e911-43f1-8b08-763891fa6038) FTR Configs #<span></span>39 / Serverless Common UI - Management Index Management Index Templates Create index template Creates index template
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/298961#0196a52c-a1c1-4f94-94cb-276ad2adc1d7) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/298961/jobs/0196a52c-a1c1-4f94-94cb-276ad2adc1d7/artifacts/0196a54a-fde1-403f-9b26-a628fbed4a9c) FTR Configs #<span></span>8 / Serverless Index Management APIs Index templates get one with logs-*-* index pattern "before all" hook for "returns logsdb index mode if logsdb.enabled setting is true"
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/298961#0196a52c-a222-467c-988f-24e8dfb4cc5d) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/298961/jobs/0196a52c-a222-467c-988f-24e8dfb4cc5d/artifacts/0196a54a-9524-411a-af28-845222a29c3c) FTR Configs #<span></span>121 / Serverless Index Management APIs Index templates get one with logs-*-* index pattern "before all" hook for "returns logsdb index mode if logsdb.enabled setting is true"
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/298961#0196a52c-a226-45da-8bfd-deb580b80a5c) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/298961/jobs/0196a52c-a226-45da-8bfd-deb580b80a5c/artifacts/0196a54d-bc03-4ca6-a0b3-b10ccda31c99) FTR Configs #<span></span>126 / Serverless Index Management APIs Index templates get one with logs-*-* index pattern "before all" hook for "returns logsdb index mode if logsdb.enabled setting is true"

### History
* :broken_heart: [Build #298915](https://buildkite.com/elastic/kibana-pull-request/builds/298915) failed 8fecca704cec7e77a248e486d408d3e3fdf3c40b

cc @SoniaSanzV

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/298961","number":298961,"commit":"dffba95833e0599831ba063560dbcfb4bd8bcc40"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/298915","number":298915,"commit":"8fecca704cec7e77a248e486d408d3e3fdf3c40b"}],"number":298961}
buildkite-pr-comment-->

SoniaSanzV committed:
Fix bug on save

elasticmachine commented:
## :yellow_heart: Build succeeded, but was flaky
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/299290)
* Commit: a48c47cffaf62ca9e7c6c6071b7c8e9de2041c2a

### Failed CI Steps
* [Jest Tests #17](https://buildkite.com/elastic/kibana-pull-request/builds/299290#0196a9d3-d19f-492a-80d0-b4205d27806c)

### Test Failures
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/299290#0196a9d3-d19f-492a-80d0-b4205d27806c) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/299290/jobs/0196a9d3-d19f-492a-80d0-b4205d27806c/artifacts/0196a9fa-f6d5-4c6b-b2c4-7e64f693caf6) Jest Tests #<span></span>17 / Connectors the connector is changed successfully

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Page load bundle
> Size of the bundles that are downloaded on every page load. Target size is below 100kb


| id | [before](https://github.com/elastic/kibana/commit/f5afb22941f6008db2f22c1b6fda781e6f9c7e4c) | [after](https://github.com/elastic/kibana/commit/a48c47cffaf62ca9e7c6c6071b7c8e9de2041c2a) | diff |
| --- | --- | --- | --- |
| `indexManagement` | 31.7KB | 31.7KB | +32.0B |



### History
* :broken_heart: [Build #299089](https://buildkite.com/elastic/kibana-pull-request/builds/299089) failed 36187b904e8f139c5a3af4f0b2900b64dd2aad98
* :broken_heart: [Build #298961](https://buildkite.com/elastic/kibana-pull-request/builds/298961) failed dffba95833e0599831ba063560dbcfb4bd8bcc40
* :broken_heart: [Build #298915](https://buildkite.com/elastic/kibana-pull-request/builds/298915) failed 8fecca704cec7e77a248e486d408d3e3fdf3c40b

cc @SoniaSanzV

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/299290","number":299290,"commit":"a48c47cffaf62ca9e7c6c6071b7c8e9de2041c2a"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/299089","number":299089,"commit":"36187b904e8f139c5a3af4f0b2900b64dd2aad98"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/298961","number":298961,"commit":"dffba95833e0599831ba063560dbcfb4bd8bcc40"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/298915","number":298915,"commit":"8fecca704cec7e77a248e486d408d3e3fdf3c40b"}],"number":299290}
buildkite-pr-comment-->

sabarasaba commented on x-pack/platform/plugins/shared/index_management/server/routes/api/component_templates/schema_validation.ts:
```diff
@@ -19,6 +19,7 @@ export const componentTemplateSchema = schema.object({
         data_retention: schema.maybe(schema.string()),
       })
     ),
+    data_stream_options: schema.maybe(schema.object({}, { unknowns: 'allow' })),
```
nit: can we add a small comment here on why we are allowing unknowns for the time being?

sabarasaba commented on x-pack/platform/plugins/shared/index_management/server/routes/api/component_templates/register_update_route.ts:
```diff
@@ -41,11 +41,32 @@ export const registerUpdateRoute = ({
+          method: 'GET',
+          path: `/_component_template/${name}`,
+        });
+        // TBD: Replace with the following when the client includes data_stream_options in ClusterComponentTemplateSummary
```
can we point to an ticket here? Will be easier to reference later on

sabarasaba commented on x-pack/platform/plugins/shared/index_management/server/routes/api/templates/lib.ts:
```diff
@@ -55,3 +57,34 @@ export const saveTemplate = async ({
+    method: 'GET',
+    path: `/_index_template/${name}`,
+  });
+  // TBD: Replace with the following when the client includes data_stream_options in IndicesIndexTemplateSummary
```
can we point to an ticket here? Will be easier to reference later on

sabarasaba commented on x-pack/platform/plugins/shared/index_management/common/lib/template_serialization.ts:
```diff
@@ -23,7 +23,10 @@ import {
-export function serializeTemplate(templateDeserialized: TemplateDeserialized): TemplateSerialized {
+export function serializeTemplate(
+  templateDeserialized: TemplateDeserialized,
+  dataStreamOptions?: object | undefined
```
nit: would it make sense to make this type slightly more specific than just object?

```typescript
export interface DataStreamOptions {
  failure_store?: {
    enabled: boolean;
  };
  [key: string]: unknown;
}
```

sabarasaba reviewed (commented):
Thanks for having a look @SoniaSanzV! I've checked the code and all looks good, left some very minor comments.

One thing I did noticed is that I was expecting to see the datastreamsoptions in the review details section, given that is included in the request that is being made to ES:

![Screenshot 2025-05-08 at 09 20 24](https://github.com/user-attachments/assets/daa8dab3-4916-4c31-a924-9c2ec39cf502)

SoniaSanzV committed:
Add data_strams_options in request preview

SoniaSanzV committed:
Merge branch 'main' into index_management/persist-data_stream_options-templates

SoniaSanzV committed:
Add component template serialization test

SoniaSanzV commented:
Thank you for the review @sabarasaba! And great catch on the request previews, I hadn't taken them into account! I've addressed your feedback in c9e91a2652a74e35e6c7e2f493fa3ed1c46f6fe9

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/299993)
* Commit: 2c43fa6ac28effd86183b363080edcf4ca5c912d

### Failed CI Steps
* [Linting](https://buildkite.com/elastic/kibana-pull-request/builds/299993#0196b459-8b36-4829-b213-38d13c943b0d)

### History
* :yellow_heart: [Build #299290](https://buildkite.com/elastic/kibana-pull-request/builds/299290) was flaky a48c47cffaf62ca9e7c6c6071b7c8e9de2041c2a
* :broken_heart: [Build #299089](https://buildkite.com/elastic/kibana-pull-request/builds/299089) failed 36187b904e8f139c5a3af4f0b2900b64dd2aad98
* :broken_heart: [Build #298961](https://buildkite.com/elastic/kibana-pull-request/builds/298961) failed dffba95833e0599831ba063560dbcfb4bd8bcc40
* :broken_heart: [Build #298915](https://buildkite.com/elastic/kibana-pull-request/builds/298915) failed 8fecca704cec7e77a248e486d408d3e3fdf3c40b

cc @SoniaSanzV

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/299993","number":299993,"commit":"2c43fa6ac28effd86183b363080edcf4ca5c912d"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/299290","number":299290,"commit":"a48c47cffaf62ca9e7c6c6071b7c8e9de2041c2a"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/299089","number":299089,"commit":"36187b904e8f139c5a3af4f0b2900b64dd2aad98"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/298961","number":298961,"commit":"dffba95833e0599831ba063560dbcfb4bd8bcc40"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/298915","number":298915,"commit":"8fecca704cec7e77a248e486d408d3e3fdf3c40b"}],"number":299993}
buildkite-pr-comment-->

SoniaSanzV committed:
Remove console log

elasticmachine commented:
## :yellow_heart: Build succeeded, but was flaky
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/300030)
* Commit: 277d161dd39f60f41ce370e78be0e4857e8fd8d9

### Failed CI Steps
* [FTR Configs #121](https://buildkite.com/elastic/kibana-pull-request/builds/300030#0196b4a7-929e-47fa-bff0-cc4d7db74689)

### Test Failures
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/300030#0196b4a7-929e-47fa-bff0-cc4d7db74689) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/300030/jobs/0196b4a7-929e-47fa-bff0-cc4d7db74689/artifacts/0196b4c6-2720-4176-bbda-8ad2efeec95c) FTR Configs #<span></span>121 / Cloud Security Posture Vulnerabilities Page - DataTable "before all" hook in "Vulnerabilities Page - DataTable"

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/1dd80c9c51c3c47f9f5ab0f059ccc3691883fd1f) | [after](https://github.com/elastic/kibana/commit/277d161dd39f60f41ce370e78be0e4857e8fd8d9) | diff |
| --- | --- | --- | --- |
| `indexManagement` | 683.2KB | 683.4KB | +208.0B |

#### Public APIs missing exports
> Total count of every type that is part of your API that should be exported but is not. This will cause broken links in the API documentation system. Target amount is 0. `Run node scripts/build_api_docs --plugin [yourplugin] --stats exports` for more detailed information.


| id | [before](https://github.com/elastic/kibana/commit/1dd80c9c51c3c47f9f5ab0f059ccc3691883fd1f) | [after](https://github.com/elastic/kibana/commit/277d161dd39f60f41ce370e78be0e4857e8fd8d9) | diff |
| --- | --- | --- | --- |
| `indexManagement` | 1 | 2 | +1 |

#### Page load bundle
> Size of the bundles that are downloaded on every page load. Target size is below 100kb


| id | [before](https://github.com/elastic/kibana/commit/1dd80c9c51c3c47f9f5ab0f059ccc3691883fd1f) | [after](https://github.com/elastic/kibana/commit/277d161dd39f60f41ce370e78be0e4857e8fd8d9) | diff |
| --- | --- | --- | --- |
| `indexManagement` | 31.7KB | 31.7KB | +69.0B |



### History
* :broken_heart: [Build #299993](https://buildkite.com/elastic/kibana-pull-request/builds/299993) failed 2c43fa6ac28effd86183b363080edcf4ca5c912d
* :yellow_heart: [Build #299290](https://buildkite.com/elastic/kibana-pull-request/builds/299290) was flaky a48c47cffaf62ca9e7c6c6071b7c8e9de2041c2a
* :broken_heart: [Build #299089](https://buildkite.com/elastic/kibana-pull-request/builds/299089) failed 36187b904e8f139c5a3af4f0b2900b64dd2aad98
* :broken_heart: [Build #298961](https://buildkite.com/elastic/kibana-pull-request/builds/298961) failed dffba95833e0599831ba063560dbcfb4bd8bcc40
* :broken_heart: [Build #298915](https://buildkite.com/elastic/kibana-pull-request/builds/298915) failed 8fecca704cec7e77a248e486d408d3e3fdf3c40b

cc @SoniaSanzV

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/300030","number":300030,"commit":"277d161dd39f60f41ce370e78be0e4857e8fd8d9"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/299993","number":299993,"commit":"2c43fa6ac28effd86183b363080edcf4ca5c912d"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/299290","number":299290,"commit":"a48c47cffaf62ca9e7c6c6071b7c8e9de2041c2a"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/299089","number":299089,"commit":"36187b904e8f139c5a3af4f0b2900b64dd2aad98"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/298961","number":298961,"commit":"dffba95833e0599831ba063560dbcfb4bd8bcc40"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/298915","number":298915,"commit":"8fecca704cec7e77a248e486d408d3e3fdf3c40b"}],"number":300030}
buildkite-pr-comment-->

SoniaSanzV committed:
Add data_stream_options to Preview index template

SoniaSanzV committed:
Merge branch 'main' into index_management/persist-data_stream_options-templates

sabarasaba reviewed (approved):
latest changes lgtm 🚀
